src/main/java/graphql/validation/RulesVisitor.java

```diff
@@ -25,41 +24,33 @@
 import graphql.language.VariableReference;
 
 @Internal
+@SuppressWarnings("rawtypes")
 public class RulesVisitor implements DocumentVisitor {
-
-    private final ImmutableList<AbstractRule> rules;
     private final ValidationContext validationContext;
-    private boolean subVisitor;
-    private final List<AbstractRule> rulesVisitingFragmentSpreads = new ArrayList<>();
-    private final Map<Node, List<AbstractRule>> rulesToSkipByUntilNode = new IdentityHashMap<>();
-    private final Set<AbstractRule> rulesToSkip = new LinkedHashSet<>();
+    private int fragmentSpreadVisitDepth;
+    private final Stack<Node> popRulesStackOnLeave = new Stack<>();
+    private final Stack<ImmutableList<AbstractRule>> rulesStack = new Stack<>();
```

Comment:
this does a similar job as `rulesToSkip`. Whereas rulesToSkip describes muted rules that should be removed from iteration, `rulesStack` describes the subset of rules that should be iterated. Rules are muted by pushing a new list on the stack that does not contain the muted rules.